Fixed touch in menues behavior

euquiq opened the pull request:
Reference:

https://github.com/eried/portapack-mayhem/issues/69

(I managed to create another clean PR!!!!)

eried commented:
Thanks a lot @euquiq, the only non clean part is firmware/common/ui_widget.cpp, if you notice most of lines have just indentation switch (I'm not completely sure but it is kinda an assholes move for the original author of that line: i.e. it switches the authority to you and looks untidy in the commit). In VS code there is a mark on the side signaling changes on the line vs git and you can actually fix those details before creating the commit inside vs code itself (just reject the new tabbed lines)

euquiq commented:
> Thanks a lot @euquiq, the only non clean part is firmware/common/ui_widget.cpp, if you notice most of lines have just indentation switch (I'm not completely sure but it is kinda an assholes move for the original author of that line: i.e. it switches the authority to you and looks untidy in the commit). In VS code there is a mark on the side signaling changes on the line vs git and you can actually fix those details before creating the commit inside vs code itself (just reject the new tabbed lines)

Uhmm I may not be exactly understanding what you are commenting. JIC I edited the ui_widget.cpp and cleaned an extra space at the beginning of each commented line, as also deleted an extra empty line inside the commented code ? If it is not what you are referring to, please explain

eried commented:
Oh I think it was all ok! Sorry. I was checking in the phone and didn't saw the comments (maybe just a block comment would have been better :D ) but thanks and sorry

eried commented:
Did you tested this on the main menu? If you tap something and then tap back sometimes the second tap only selects the item instead of entering the submenu. Example:

1. Turn device on
2. Tap scanner

Selected item changes but nothing else happens.

euquiq commented:
> Did you tested this on the main menu? If you tap something and then tap back sometimes the second tap only selects the item instead of entering the submenu. Example:
> 
> 1. Turn device on
> 2. Tap scanner
> 
> Selected item changes but nothing else happens.

Yes, I tested it, and It works as intended. I do not get the behavior you are getting!

Just to be sure, again, I just recompiled it using the branch which is your code plus the PR, and it works correctly.

euquiq commented:
Hold the horses ... there MIGHT be something weirder going on in here. I get "sometimes" (when trying your firmware) a behavior similar to what you describe, but with the DEBUG menu option.

Now the weirdest thing: If i press on the debug button, and while pressing, I move the finger away the debug button, it works.

This seems to be something more electrical and even a deficiency on my H1 (Chinese crappy touchscreen). 

I will flash now my version of the firmware (The one that includes the ".." back buttons and try the same with it. 

I will investigate this. Weird!

euquiq commented:
There are some problems with this PR. 

One is: the current behavior of the button menues is "by design" as it is. It has no provision in its code to "memorize" last pressed menu option, when using the on_touch() function. So it is my understanding that the issue mentioned in https://github.com/eried/portapack-mayhem/issues/69  it is not a "BUG". But, maybe, an enhancement. Which is not 100% clear, either.

Another problem is regarding my fix: In my portapack H1 it is working OK, but I had to lower the **r_touch_threshold** on line 215 of `/firmware/application/touch.hpp` from its default value of 640 down to 320. 

Which made me investigate (but I know very little of C++). 

This PR moves the on_touch event processing, from it's original place (inside ui_widgets.cpp) into ui_btngrid.cpp, by overriding it.

My conclusion is that this introduces a change on timing, which incresases the noise on the touchscreen reading. Hence the problem detecting pressed buttons, unless the r_touch_threshold is lowered.

This may be OK for some people, but not ok for others. It definitely is not optimal.

I am looking into a better way of resolving this ... but also I think the issue itself needs to be re-considered. Is it an enhancement ? Is it a bug ? is it worth it ?
